### PR TITLE
Work towards shared generics support in IDynamicIntefaceCastable

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
This is enough to get all our test to pass.

The rest is tracked in https://github.com/dotnet/runtime/issues/72909.

It's a bunch of work and at this point I don't know how much we need it.

### CODE REVIEW

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/SharedCodeHelpers.cs [hunk 1]
```diff
@@ -15,5 +15,11 @@ internal static class SharedCodeHelpers
             Debug.Assert(interfaceIndex <= pType->NumInterfaces);
             return pType->InterfaceMap[interfaceIndex].InterfaceType;
         }
+
+        public static unsafe MethodTable* GetCurrentSharedThunkContext()
+        {
+            // TODO: We should return the current context from the ThunkPool
```
Comment (jkotas) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/SharedCodeHelpers.cs:21]: Add a link to the issue you have opened? 

Maybe also add assert? Or is it actually exercised by the tests and we are just getting lucky that nothing is accessing the context?

**Review by jkotas — APPROVED** [review]

**Comment by jkotas** [comment]
> It's unclear to me if anyone uses IDynIntfCast with generics.

I think cswinrt is going to use it with generics. However, I am not sure whether it is going to hit this specific case with missing context.

**Comment by MichalStrehovsky** [comment]
> I think cswinrt is going to use it with generics. However, I am not sure whether it is going to hit this specific case with missing context.

I tried CsWinRT but at this point it's AOT unfriendly beyond any workarounds.

One thing is all the `MakeGenericType` and `MakeGenericMethod` calls and string concatenations passed to `GetType`. Those can be worked around with RD.XML. It's not pretty but it's workable.

But then there's "[declaring a delegate type requires writing a whole new line in C# so let's ask Expressions to reflection-emit a new delegate type at runtime for us instead](https://github.com/microsoft/CsWinRT/blob/dcec6e0f71bb0b9eeeec6cb25aeee20eab61744f/src/WinRT.Runtime/Projections/IReadOnlyList.net5.cs#L209-L210)".

I miss MCG.

**Comment by MichalStrehovsky** [comment]
Let's not merge yet, there's a bug.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/SharedCodeHelpers.cs (hunk 1 above)
Comment (MichalStrehovsky) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/SharedCodeHelpers.cs:21]: It is exercised by one of the tests, but the test doesn't do anything with the context.

**Comment by AaronRobinsonMSFT** [comment]
>I miss MCG.

Said no one ever :-)

**Pull request merged**